### PR TITLE
major fixes

### PR DESCRIPTION
**Pull request by henriqueblobato**

### CODE REVIEW

File: config.ini
Comment (CRAKZOR) [inline config.ini:1]: line 10, li_at might be personal

File: config.ini [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+gpt_preamble = Craft an engaging LinkedIn post. My goal is to share a short perspective on recent industry news that would be of value to my network. To help you generate a relevant and resonating post, I provided a short bio about myself. Analyze the scraped text from latest happenings in the industry. I'd like the post to be concise, informative, creative, and representative of my professional persona. Include emoticons and do not include the source. Use two enters spaces between every paragraph. Given this, please help me construct a LinkedIn post that leverages the information effectively and stands out to my audience. Focus on only one piece of news.
+gpt_token_limit = 250
+scrape_char_limit = 800
+open_ai_api_key = xxxxxxxxxx
+
+[cookies]
+JSESSIONID = ajax:123123123123123
+li_at = AQEDAUVnqTMD-K9NAAABiZkWDgYAAAGJvS6SBk0AuKFo1f6DKa-TTxMxkfApY1CkY534-Hh2P9Gjxu-BYW3os-I40toem61Br_EOdOOxfWls2PRvev5b8EamRK1IhQoYSjUnr8j_Q8Qz5fRa3pbwHuOf
```
Comment (CRAKZOR) [inline config.ini:10]: Might be personal information

File: main.py [hunk 2]
```diff
@@ -1,109 +1,122 @@
+import configparser
```
Comment (CRAKZOR) [inline main.py:1]: explain why this is better than JSON config

File: scraper.py [hunk 3]
```diff
@@ -17,22 +17,16 @@ def fetch_content(self):
     def parse(self, content):
         soup = BeautifulSoup(content, 'html.parser')
 
-        # Removing undesired elements like navigation bars and headers
-        # This is a generic example; adjust according to the specific website
-        for element in soup.find_all(["nav", "header"]):  # You can also look for elements by their class or ID
```
Comment (CRAKZOR) [inline scraper.py:22]: Why not remove nav and header html elements?

File: scraper.py [hunk 4]
```diff
@@ -17,22 +17,16 @@ def fetch_content(self):
...
 
-        # Split the lines, remove leading and trailing whitespace
         lines = (line.strip() for line in all_text.splitlines())
 
-        # Remove empty lines, and multi-space and join
-        all_text_clean = re.sub( r'\s+', ' ', ( ' '.join( line for line in lines if line ) ) )
-
-        return all_text_clean[0:self.character_limit]
+        all_text_clean = re.sub(r'\s+', ' ', (' '.join(line for line in lines if line)))
+        all_text_clean = all_text_clean.replace("  ", " ").replace('\\n', '')
```
Comment (CRAKZOR) [inline scraper.py:31]: This would happen from .replace('\\n', '')

this
example
output

thisexampleoutput

File: main.py [hunk 5]
```diff
@@ -1,109 +1,122 @@
...
-        "accept-language":          "en-US,en;q=0.9",
-        "content-type":             "application/json; charset=UTF-8",
-        "csrf-token":               cookies["JSESSIONID"],
-        "referrer-policy":          "strict-origin-when-cross-origin, strict-origin-when-cross-origin",
-        "origin":                   "https://www.linkedin.com",
-        "Referrer":                 "https://www.linkedin.com/feed/",
-        "Referrer-Policy":          "strict-origin-when-cross-origin, strict-origin-when-cross-origin",
-        "cookie":                   cookie_value
-        # ... other headers ...
+        "accept": "application/vnd.linkedin.normalized+json+2.1",
```
Comment (CRAKZOR) [inline main.py:78]: less prettier.

**Review by CRAKZOR — COMMENTED** [review]
I appreciate the changes. I need help understanding the reasoning marked by inline comments.

**Comment by henriqueblobato** [comment]
@CRAKZOR 
Hey! First of all thank you for the amazing project an idea! As I had too many changes on this PR I decided to close it, and keep it as a single fork.
At my version Im doing a post with images, I already got the requests flow coming from the api, we need the urn provided on the first request to upload the binary
Here we request for a upload urn number
```
payload = {
      "mediaUploadType": "IMAGE_SHARING",
      "fileSize": image_size,
      "filename": f"{str(uuid.uuid4()).replace('-', '')}.png"
    }
    ```
The response should be inside the response as
```
{
   "value":
          {"urn":"urn:li:digitalmediaAsset:D4D22AQEg0MssCqr03Q",
....
```
But Im still getting a 400, can you help me out? Thank you!

**Comment by CRAKZOR** [comment]
Great idea! I got it working. 

image

Check my code and compare to your fork what is missing. 

main()
```python
    fname = "burndown.png"
    fbinary, fsize, ftype = get_file_data(path.join("images", fname), protocol="rb", incl_meta=True)

    payload = {
        "mediaUploadType": "IMAGE_SHARING",
        "fileSize": fsize,
        "filename": fname
    }

    urn = post_linkedin_file( json.dumps(payload), cookies, file=fbinary, content_type=ftype )

    media = [
        # {category: "IMAGE", mediaUrn: "urn:li:digitalmediaAsset:D5622AQGlR8tHW6lD-Q", tapTargets: []}
        { "category": MEDIA_CATEGORY.IMAGE.name, "mediaUrn": urn, "tapTargets":[] }
    ]

    payload = {
        "visibleToConnectionsOnly": False,
        "externalAudienceProviders": [],
        "commentaryV2": {
            "text": gpt_res,
            "attributes": []
        },
        "origin": "FEED",
        "allowedCommentersScope": "ALL",
        "postState": "PUBLISHED",
        "media": media
    }

    post_linkedin( json.dumps(payload), cookies )
```

I duplicated post_linkedin and made changes to handle single file upload. 
```python
def post_linkedin_file(payload, cookies, file=None, content_type=None):

    if not file:
        return

    cookie_value = "li_at=%s; JSESSIONID=\"%s\"" % (cookies["li_at"], cookies["JSESSIONID"])

    headers = {
        "User-Agent":               "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3",
        "accept":                  "application/vnd.linkedin.normalized+json+2.1",
        "accept-language":          "en-US,en;q=0.9",
        "content-type":             "application/json; charset=UTF-8",
        "csrf-token":               cookies["JSESSIONID"],
        "referrer-policy":          "strict-origin-when-cross-origin, strict-origin-when-cross-origin",
        "origin":                   "https://www.linkedin.com",
        "referrer":                 "https://www.linkedin.com/feed/",
        "cookie":                   cookie_value
        # ... other headers ...
    }

    post_endpoint = "https://www.linkedin.com/voyager/api/voyagerVideoDashMediaUploadMetadata?action=upload"

    try:
        response = requests.post(post_endpoint, headers=headers, data=payload)
        response.raise_for_status()

        data                         = response.json()["data"]["value"]

        upload_endpoint              = data["singleUploadUrl"]
        headers["media-type-family"] = data["singleUploadHeaders"]["media-type-family"]
        headers["content-type"]      = content_type

        response = requests.put(upload_endpoint, headers=headers, data=file )
        # print(response.request.headers)
        # print(response.request.body)
        # raise ValueError("poop")
        # response.raise_for_status()

        # success.
        return data["urn"]

    except requests.exceptions.RequestException as e:
        # Handle exception
        custom_print(f"Error posting to LinkedIn: {e}")
        custom_print(response.text)
```

Important:
I forgot to include the user agent in header. requests lib automatically sends 'python script' as the UA. just add this to header like above.

```python
    headers = {
        "User-Agent":               "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3",
        "accept":                  "application/vnd.linkedin.normalized+json+2.1",
        "accept-language":          "en-US,en;q=0.9",
        ...
```

**Comment by CRAKZOR** [comment]
it's in 83cc85183d4c0abde79d57ef8cd1c77c71e0b105. I just don't have a use-case for it yet.

**Comment by henriqueblobato** [comment]
Oh man thats amazing! I need to merge with you so we can start working together again, let me do it another PR so we can manage all the changes.
Also my next move was migrate to class based objects, and I see that you are already moving forward to that! Great news!

**Pull request closed without merging**